### ISSUE TITLE
tcp connections taking too long to close

### ISSUE THREAD

**Issue by dheeraj-varshney** [nodejs/undici#1135]
## Bug Description

We recently migrated our http client from `request-promise-native` to `undici` on our GraphQL server. We have observed significant performance gains :tada:.

- Our GraphQL server connects through upstream micro-services through HTTP connection.
- After this migration we observed that connections between GraphQL and upstream services are not draining on regular interval.
- Upstream services are deployed using blue green deployments.
- Even after we deploy new stack of upstream service, there are connection open on older stack for more than 1.5 hrs.

Is there a config to specify to drain the TCP connections on certain intervals

### Additional context
Sample Code-
```
import {IncomingHttpHeaders} from 'http'
import {request} from 'undici'

export const nativePostRequest = async (
  url: string,
  headers: IncomingHttpHeaders,
  body: any,
  extra: string
) => {
  const bodyString = JSON.stringify(body)
  return await request(url, {
    method: 'POST',
    headers: headers,
    body: bodyString,
    bodyTimeout: config.httpTimeout
  })
}
```

**Comment by ronag** [comment]
I think this is more of server thing. undici follows the keep alive hints that the server provides.

**Comment by ronag** [comment]
You can try setting the `maxKeepAliveTimeout`

**Comment by dheeraj-varshney** [comment]
@ronag Both these fields doesn't exist in the options for request method. How do you suggest I should use these?

**Comment by mcollina** [comment]
These options are set on the Agent, see https://undici.nodejs.org/#/docs/api/Agent.

I think we should provide some sample configs for these. @ronag We might even think of lowering the default keepalivetimeout to 1s or 500ms.

**Comment by smehta91** [comment]
@mcollina @ronag Are we doing any kind of DNS caching internally?

**Comment by ronag** [comment]
Nope

**Comment by dheeraj-varshney** [comment]
@mcollina @ronag I tried with different implementation. I switched to creating new Client and then setting `keepAliveMaxTimeout`. But there is a huge performance degradation via this implementation. 

```
const httpClient = {
  'xyz': new Client(`http://localhost:3000`, {
    keepAliveMaxTimeout: config.httpTimeout
  })
}

export const nativePostRequest = async (
  url: string,
  headers: IncomingHttpHeaders,
  body: any
) => {
  const bodyString = JSON.stringify(body)
  const client = httpClient['xyz']

  return client.request({
    path: url
    method: 'POST',
    headers: headers,
    body: bodyString,
    bodyTimeout: config.httpTimeout
  })
}
```

The latency increases drastically with this implementation.

**Comment by mcollina** [comment]
You should be able to do:

```js
new Agent({
  keepAliveMaxTimeout: config.httpTimeout
})
```

**Comment by blueonrails** [comment]
Hi there,

I am testing `undici` to see if we can improve performance at a high traffic service at the company I am working for. 

At the moment we are using [agentkeepalive](https://github.com/node-modules/agentkeepalive) with some custom fine tunning. 

Testing undici I observed when incommig traffic increase a lot, the Agent creates a lot of tcp connections (maxRequestsPerClient is not setted so there is no limit) after some seconds the service with a lot of open tcp connections begins to response very slow, the service event loop lag increase a lot and start to timeout some requests. After some seconds later those tcp connections recently opened start to close because keepAliveMaxTimeout o server hint is reached, the service is gradually recovering afterwards.

In this case if I set a max of clients (maxRequestsPerClient) the service will scale badly when the traffic increase. Inspired by `agentkeepalive` maybe is interesting be able to set a limit to the open connections. Thus, `undici` can quickly create all the necessary clients, but it will not keep them alive for a long time.

Right now I am studying how to make this limit by configuring pipelining=0 when a specific amount of clients is reached. PoolStats is really coming in handy.

By the way, I have been reviewing the internal source code on undici and it looks very cool. 👏 👏 👏 

Thanks for reading me

**Comment by mcollina** [comment]
@blueonrails you can set the `connections` option to limit the maximum amount of sockets to allocate. This is likely needed to avoid the problems you are facing: there is no point to increase parallelism over a certain threshold.

**Comment by blueonrails** [comment]
@mcollina I totally agree with you in that sentence. On the other hand, the need I see is different, maybe it is more interesting to put a limit to the free connections than to the whole set of connections. Or even to both, it depends on how you are using the library. 

For example Go does support that in its native client. See: `MaxIdleConns` on https://pkg.go.dev/net/http.

Anyway, thanks a lot for the help and the lib.

**Comment by mcollina** [comment]
We use a different pattern: you can set the maximum amount of time a connection is kept alive. Set this small enough and you are going to achieve your goalz

**Comment by metcoder95** [comment]
See: https://undici.nodejs.org/#/docs/api/Client?id=parameter-clientoptions
